### PR TITLE
[Issue 7519] Adds support for multiple filtered/unfiltered aggregations with GROUP BY

### LINKED ISSUE #7519: FILTER Clause Support
FILTER clause (SELECT COUNT(*) FILTER (WHERE foo > bar) FROM fooBar) allows complex reports to be generated efficiently without recomputation of the core predicate and rescans of the original data set.

https://docs.google.com/document/d/1qJzG1CmyVZpLN6rdI8V0zUWEaIjQ86hDL9XwrcPKcTs/edit?usp=sharing

### PR DESCRIPTION
**Pull request by egalpin**
Following @atris's awesome work[1] to support FILTER expressions[2][3], this PR aims to introduce support for FILTER expressions simultaneously with GROUP BY. The same swim-lane pattern established in the initial FILTER expression PR[1] is re-used/shared here by groupBy processing.

Screen Shot 2022-12-16 at 17 19 04

Fixes #7519

[1] https://github.com/apache/pinot/pull/7916
[2] https://github.com/apache/pinot/issues/7519
[3] https://docs.google.com/document/d/1ZM-2c0jJkbeJ61m8sJF0qj19t5UYLhnTFvIAz-HCJmk

### CODE REVIEW

**Comment by egalpin** [comment]
Will work on the checkstyle violations.  Has there ever been consideration of using Spotless[1] which can both check as well as autofix linting issues?  I think pinot-style specifically is not supported, but perhaps support could be added?  Auto-fixing would be a huge win.

[1] https://github.com/diffplug/spotless

**Comment by egalpin** [comment]
@Jackie-Jiang this is ready for review 😊

File: pinot-core/src/test/java/org/apache/pinot/core/query/aggregation/groupby/GroupByTrimTest.java [hunk 1]
```diff
@@ -130,7 +129,7 @@ void testGroupByTrim(QueryContext queryContext, int minSegmentGroupTrimSize, int
     // Extract the execution result
     List<Pair<Double, Double>> extractedResult = extractTestResult(resultsBlock.getTable());
 
-    assertEquals(extractedResult, expectedResult);
+    Assert.assertEquals(extractedResult, expectedResult);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/core/query/aggregation/groupby/GroupByTrimTest.java:132]: (nit) We usually use static import for `Assert` in test

File: pinot-core/src/test/java/org/apache/pinot/queries/FilteredAggregationsTest.java [hunk 2]
```diff
@@ -335,10 +335,13 @@ public void testMixedAggregationsOfSameType() {
...
   public void testGroupBySupport() {
-    String filterQuery = "SELECT MIN(INT_COL) FILTER(WHERE NO_INDEX_COL > 2), MAX(INT_COL) FILTER(WHERE INT_COL > 2) "
-        + "FROM MyTable WHERE INT_COL < 1000 GROUP BY INT_COL";
-    getBrokerResponse(filterQuery);
+    String filterQuery =
+        "SELECT SUM(INT_COL), SUM(INT_COL) FILTER(WHERE INT_COL > 25000) AS total_sum FROM MyTable GROUP BY INT_COL";
+    String nonFilterQuery =
+        "SELECT SUM(INT_COL), SUM(CASE WHEN INT_COL > 25000 THEN INT_COL ELSE 0 END) AS total_sum FROM MyTable GROUP "
+            + "BY INT_COL";
+    testQuery(filterQuery, nonFilterQuery);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/FilteredAggregationsTest.java:345]: Let's add some more tests to guarantee that it works as expected

File: pinot-core/src/main/java/org/apache/pinot/core/plan/GroupByPlanNode.java [hunk 3]
```diff
@@ -50,10 +54,35 @@ public GroupByPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
...
     assert _queryContext.getGroupByExpressions() != null;
 
+    if (_queryContext.hasFilteredAggregations()) {
+      return filteredGroupByPlan();
+    }
+
+    return nonFilteredGroupByPlan();
+  }
+
+  private FilteredGroupByOperator filteredGroupByPlan() {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/GroupByPlanNode.java:68]: (minor) Rename to `buildFilteredGroupByPlan()` and `buildNonFilteredGroupByPlan()`

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionUtils.java [hunk 4]
```diff
@@ -71,11 +82,13 @@ public static AggregationFunctionColumnPair getAggregationFunctionColumnPair(
    * <p>NOTE: We don't need to consider order-by columns here as the ordering is only allowed for aggregation functions
    *          or group-by expressions.
    */
-  public static Set<ExpressionContext> collectExpressionsToTransform(AggregationFunction[] aggregationFunctions,
-      @Nullable ExpressionContext[] groupByExpressions) {
+  public static Set<ExpressionContext> collectExpressionsToTransform(
+      @Nullable AggregationFunction[] aggregationFunctions, @Nullable ExpressionContext[] groupByExpressions) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionUtils.java:86]: `aggregationFunctions` should always be non-null

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java [hunk 5]
```diff
@@ -52,24 +52,39 @@ public class DefaultGroupByExecutor implements GroupByExecutor {
 
   protected final AggregationFunction[] _aggregationFunctions;
   protected final boolean _nullHandlingEnabled;
-  protected final GroupKeyGenerator _groupKeyGenerator;
-  protected final GroupByResultHolder[] _groupByResultHolders;
   protected final boolean _hasMVGroupByExpression;
   protected final int[] _svGroupKeys;
   protected final int[][] _mvGroupKeys;
+  protected GroupKeyGenerator _groupKeyGenerator;
+  protected GroupByResultHolder[] _groupByResultHolders;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java:59]: These 2 variables can still be `final`

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java [hunk 6]
```diff
@@ -52,24 +52,39 @@ public class DefaultGroupByExecutor implements GroupByExecutor {
...
+
+  public DefaultGroupByExecutor(QueryContext queryContext, ExpressionContext[] groupByExpressions,
+      TransformOperator transformOperator) {
+
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java:63]: (nit) Remove empty line

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java [hunk 7]
```diff
@@ -52,24 +52,39 @@ public class DefaultGroupByExecutor implements GroupByExecutor {
...
+   * @param aggregationFunctions Aggregation functions
    * @param groupByExpressions Array of group-by expressions
    * @param transformOperator Transform operator
    */
-  public DefaultGroupByExecutor(QueryContext queryContext, ExpressionContext[] groupByExpressions,
-      TransformOperator transformOperator) {
-    _aggregationFunctions = queryContext.getAggregationFunctions();
+  public DefaultGroupByExecutor(QueryContext queryContext, AggregationFunction[] aggregationFunctions,
+      ExpressionContext[] groupByExpressions, TransformOperator transformOperator,
+      GroupKeyGenerator groupKeyGenerator) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/groupby/DefaultGroupByExecutor.java:83]: Annotate `groupKeyGenerator` as `Nullable`

**Review by Jackie-Jiang — COMMENTED** [review]
Very well done! Mostly minor comments

File: pinot-core/src/test/java/org/apache/pinot/core/query/aggregation/groupby/GroupByTrimTest.java (hunk 1 above)
Comment (egalpin) [reply pinot-core/src/test/java/org/apache/pinot/core/query/aggregation/groupby/GroupByTrimTest.java:132]: I had changed this due to checkstyle violations:

```[WARNING] (imports) AvoidStaticImport: Using a static member import should be avoided```

I can revert back so long as checkstyle still passes

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/FilteredGroupByOperator.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    for (Pair<AggregationFunction[], TransformOperator> filteredAggregation : _aggFunctionsWithTransformOperator) {
+      TransformOperator transformOperator = filteredAggregation.getRight();
+      AggregationFunction[] filteredAggFunctions = filteredAggregation.getLeft();
+
+      // Perform aggregation group-by on all the blocks
+      DefaultGroupByExecutor groupByExecutor;
+      if (groupKeyGenerator == null) {
+        // The group key generator should be shared across all AggregationFunctions so that agg results can be
```
Comment (egalpin) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/FilteredGroupByOperator.java:125]: @Jackie-Jiang please let me know where I may have misstated the reasoning

**Comment by egalpin** [comment]
@Jackie-Jiang added more tests in https://github.com/apache/pinot/pull/10000/commits/aeebef0e7ee63337615ec894f177f989cdf00751, let me know if there are other specific cases to cover. 

This PR is ready for re-review

**Comment by egalpin** [comment]
Oops, somehow clicking `request re-review` from Jackie also removed Mayank? Removal was unintentional

**Comment by egalpin** [comment]
@Jackie-Jiang the integration test failure looks unrelated to me as the same tests are failing in the same way on other PRs / the failures don't seem to point to the changed areas of code.  Can you confirm?

**Review by Jackie-Jiang — APPROVED** [review]
LGTM!

**Comment by kishoreg** [comment]
Thanks  @egalpin for your contribution. Please add docs when you get a chance if its not done already

**Comment by egalpin** [comment]
Will do @kishoreg 👍

**Comment by ankitsultana** [comment]
@egalpin : can you update documentation as well? https://docs.pinot.apache.org/users/user-guide-query/query-syntax/supported-aggregations#filter-clause-in-aggregation

This still says the following:

```
NOTE: The FILTER clause is currently supported for aggregation-only queries, i.e., GROUP BY
is not supported.
```

**Comment by egalpin** [comment]
Thanks!  https://github.com/pinot-contrib/pinot-docs/pull/250

**Pull request merged**